### PR TITLE
Ucr date utils

### PR DESCRIPTION
**Pull request by sravfeyn**
Adds bunch of date-utilities to UCR expressions. See [notes](https://github.com/dimagi/commcare-hq/pull/10354/files#diff-ee0bc703fa9242892861d966c8a6474cR90) on docs regarding the particulars

Prep for ICDS UCRs
@nickpell cc @czue

### CODE REVIEW

File: corehq/apps/userreports/tests/test_date_expressions.py [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+            'type': 'property_name',
+            'property_name': 'dob',
+        },
+        'months_expression': months_expression
+    })
+    self.assertEqual(expected_value, expression(source_doc))
+
+
+@generate_cases([
+    ({'dob': '2015-01-20'},  date(2015, 1, 1)),
```
Comment (dimagimon) [inline corehq/apps/userreports/tests/test_date_expressions.py]: E241 multiple spaces after ','

File: corehq/apps/userreports/tests/test_date_expressions.py [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+    ({'dob': date(2015, 1, 31)},  date(2015, 1, 1)),
```
Comment (dimagimon) [inline corehq/apps/userreports/tests/test_date_expressions.py]: E241 multiple spaces after ','

File: corehq/apps/userreports/tests/test_date_expressions.py [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+        'date_expression': {
+            'type': 'property_name',
+            'property_name': 'dob',
+        },
+    })
+    self.assertEqual(expected_value, expression(source_doc))
+
+
+@generate_cases([
+    ({'dob': '2015-01-20'},  date(2015, 1, 31)),
```
Comment (dimagimon) [inline corehq/apps/userreports/tests/test_date_expressions.py]: E241 multiple spaces after ','

File: corehq/apps/userreports/tests/test_date_expressions.py [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+    ({'dob': date(2015, 2, 20)},  date(2015, 2, 28)),
```
Comment (dimagimon) [inline corehq/apps/userreports/tests/test_date_expressions.py]: E241 multiple spaces after ','

File: corehq/apps/userreports/README.md [hunk 5]
```diff
@@ -322,6 +326,58 @@ The date_expression and count_expression can be any valid expressions, or simply
...
+
+```json
+{
+    "type": "month_start_date",
+    "date_expression": {
+        "type": "property_name",
+        "property_name": "dob",
+    },
+}
+```
```
Comment (czue) [inline corehq/apps/userreports/README.md]: nice docs!

File: corehq/apps/userreports/tests/test_date_expressions.py [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+def test_diff_days_expression(self, source_doc, to_date_expression, expected_value):
+    expression = ExpressionFactory.from_spec({
+        'type': 'diff_days',
+        'from_date_expression': {
+            'type': 'property_name',
+            'property_name': 'dob',
+        },
+        'to_date_expression': to_date_expression
+    })
+    self.assertEqual(expected_value, expression(source_doc))
```
Comment (czue) [inline corehq/apps/userreports/tests/test_date_expressions.py]: nice tests!

**Comment by czue** [comment]
this is great, thanks!

**Pull request merged**